### DISCUSSION TITLE
MOA vs river (Regression)

### DISCUSSION THREAD

**Discussion by javiuni** [online-ml/river#812] [Q&A]
Hello,

I'm trying to compare the results of these two popular frameworks ([MOA](https://moa.cms.waikato.ac.nz/)) for machine learning on streaming data. In particular, I'm interested in regression models with one sample at a time.

My aim would be to demonstrate that using the same dataset and same model with both frameworks you can get the same (or very close) MAE. Do you have any suggestions on how to achieve this? Is it really possible? Have you done similar tests? 

I'm having trouble choosing a model with the same parameters present in both frameworks. My 2 best attempts have been the following, but the benchmark results are not very good:

1. For a private dataset, whose target has a standard deviation of 196. AMRules (river) MAE: 54 vs RandomAMRules (MOA) MAE: 14
2. For a reduced version [of this public dataset](https://www.kaggle.com/noqcks/aws-spot-pricing-market/home). LinearRegression with optimizer SGD (river) MAE: 0.61 vs SGD Squaredloss (MOA) MAE: 0.84

I'm looking for any suggestions on which models to compare or which public dataset to use. Is this comparison of frameworks an impossible task?

Thanks

**Comment by smastelini** [comment]
Hi @javiuni. In the past, I ran some comparisons when implementing ARF and AMRules. It's great to see comparisons like these shared with the community! Keep on this great work 📈 

My suggestion:

*Use the same comparison parameters as much as possible. `progressive_val_score` (River) and `InterleavedTestThenTrain` (MOA), although similar, have some differences. Make sure no warm start is being applied in MOA, or it will have the edge over River. Alternatively, you can perform a `for` to warm start the River models before running the progressive validation.*

PS: RandomAMRules is not equivalent to River's AMRules. The former is designed to be a base learner in ensembles of rule-based models, AFAIK. You should look for the vanilla AMRules. Also, make sure the same parameters are used, including the concept drift detectors (that can be set up in River but is kept fixed in MOA) and other hyperparameters. You could also check the linear models used in AMRules (totally configurable in River). I've used the vanilla `linear_model.LinearRegression` in River. The linear model in MOA's AMRules is hardcoded. Lastly, MOA's AMRules performs feature standardization internally (akin to `StandardScaler`). With River, the user can (and should, hahaha) choose their preferred standardization strategy. I suggest starting with `preprocessing.StandardScaler` for that matter.

Waiting to see your upcoming results! 😄
